Docker build (#42)
Ouranosinc/jupyterhub: README.md

with extra required packages, see
[`Dockerfile`](https://github.com/Ouranosinc/jupyterhub/blob/master/Dockerfile)

## Release instructions

1. Open a pull request to make desired Dockerfile configuration changes.

> 1. Install `bump-my-version`: `$ pip install -r requirements.txt`.

Reviewer: Just because I always assume that users will be at the root of the repo unless we say otherwise:

```suggestion
1. Install `bump-my-version`: `$ pip install -r .github/requirements.txt`.
```

Author: Thanks for this. I was going back and forth on having the requiremnts in the top-level or in `.github`. For here, `.github` makes the most sense.

> 1. Run `$ bump-my-version {major | minor | patch | date}`

Reviewer 2: ```suggestion
1. Run `$ bump-my-version --tag #.#.#-YYYYmmdd` or `$ bump-my-version bump date` if re-release on same upstream version
```

I don't think we can ever use the form `bump {major | minor | patch}` because that version is coming from upstream and we need to match the version from upstream.  In addition we have the build date that changes.  

Unless if we just do `bump patch`, the date field default to "now"?  But that means we have to systematically release each time upstream release, else we could jump a patch/minor version and it won't match with upstream anymore.

What happen if we need to do more than one release per day?  Can this tag format be supported `#.#.#-YYYYmmdd-{any suffix}`?

Author: I was thinking about this too. Given that Renovate manages the Jupyterhub image tag, we need to ensure that we match the same bump when preparing the release. I've never worked with Renovate and from my understanding of the docs for it, it doesn't manage dates (beyond the scope of the tool).

We do currently have the option of bumping the date (`$ bump-my-version bump date`) and we could also add a <release> modifier if we wanted ("alpha/beta/gamma", "build.1/2/3" or whatever we want). If we want to add more, I'd just add the time for simplicity (`#.#.#-YYYYmmddhhMM`).

Reviewer 2: > I've never worked with Renovate and from my understanding of the docs for it, it doesn't manage dates (beyond the scope of the tool).

Last time I checked, Renovate only update the dependencies hardcoded version, it does not handle release so it won't manage dates (which is added on top of the dependencies/upstream version). 

> We do currently have the option of bumping the date (`$ bump-my-version bump date`) and we could also add a modifier if we wanted ("alpha/beta/gamma", "build.1/2/3" or whatever we want). If we want to add more, I'd just add the time for simplicity (`#.#.#-YYYYmmddhhMM`).

Oh then add an example for adding a modifier "build.1/2/3" then.  That will solve the issue of multiple releases in a day (never happened before but I'd rather the process do not block us).

Author: Done.

>     * Be sure to match the `jupyterhub/jupyterhub` version in the tag so we know what version our image is based on.

Reviewer 2: ```suggestion
    * Be sure to match the same original upstream `jupyterhub/jupyterhub` version in the tag so we know what version our image is based on.
```

Re-include "same original upstream" to strongly emphasis we need to match with that.  Our internal versioning is basically the date string, not that version from upstream.

    * If changes do not modify the `jupyterhub/jupyterhub` image, one can simply run `$ bump-my-version bump date`.

>     * For multiple releases on the same date, one can simply run `$ bump-my-version bump build`.

Reviewer 2: Can we still do `bump-my-version --tag #.#.#-YYYYmmdd.b`?  Asking because we do not release every single time upstream release.  Or are we forced to do that now?

Recall in comment https://github.com/Ouranosinc/jupyterhub/pull/42#discussion_r3633765955, if we do not always release each time upstream release, our version won't match, ex: if upstream is now `5.6.7` but since we did not release when it was previously `5.6.6`, we are still at `5.6.5` so `bump patch` will only bring us to `5.6.6` and not `5.6.7` to match upstream.

Plz update the doc if we are forced to always release with upstream or how to handle when our version falls behind vs upstream.

Author: `$ bump-my-version bump --new-version #.#.#-YYYYmmdd.b` is allowed, if needed. I'll add that as a final option. The thing is though, that nothing is tagged until it gets pushed to `master`, so nothing is stopping you from running a sequence of bump commands if you want to be safe.

> 1. Create a tag matching the new version string (`{jupyterhub/jupyterhub version}-{YYYY}{0M}{0D}`) and push to the repository.

Reviewer 2: So we still need to manually tag and push to github after merging the PR?  Can this detect the new tag from bump-my-version and create a matching github tag?  Not blocking, can merge this PR and automate this later or not automate this at all.  This is a low publishing volume repo.  We do not systematically release for each of upstream release.

Author: The challenge here is determining when and who creates this tag.

I could confgure bump-my-version to create tags on bump, but then we might run into issues where we need to bump twice (e.g. `5.5.0-<date>` -> `5.6.1-<date>` (minor + patch)). We also absolutely do not want to create tags within a PR. 

I could potentially use a `RELEASE` file like we do elsewhere that triggers a tag creation when it changes on `main` and pushes that to the repo. That might make the most sense. GitHub has some rules around triggers caused by workflows triggering other workflows though. Will read up.

Reviewer 2: > I could confgure bump-my-version to create tags on bump, but then we might run into issues where we need to bump twice (e.g. `5.5.0-<date>` -> `5.6.1-<date>` (minor + patch)). We also absolutely do not want to create tags within a PR.

`bump-my-version` should not create the tag.  It's the github workflow, after the commit from `bump-my-version` has been pushed to `main`, like you did for birdhouse-deploy repo.  You even automated the release creation on github.

    * Be sure to match the `jupyterhub/jupyterhub` version in the tag so we know what version our image is based on.
    * If changes do not modify the `jupyterhub/jupyterhub` image, one can simply run `$ bump-my-version bump date`.
1. Merge the Pull Request to the `master` branch and switch to `master` branch.
1. Create a tag matching the new version string (`{jupyterhub/jupyterhub version}-{YYYY}{0M}{0D}`) and push to the repository.

Keep the same original upstream `jupyterhub/jupyterhub` version in the tag so

> we know what version our image is based on.

Reviewer 2: Please preserve this important info "Keep the same original upstream `jupyterhub/jupyterhub` version in the tag so we know what version our image is based on."

Author: I moved this information just above. It's still there.
